src/protocol/resp/src/parse/parser.rs

```diff
@@ -0,0 +1,494 @@
+    }
+
+    pub fn parse_simple_string(&mut self) -> ParseResult<'a, &'a [u8]> {
+        self.try_parse(|p| {
+            p.parse_literal(b"+")?;
+            p.parse_delimited_text()
+        })
+    }
+
+    pub fn parse_bulk_string(&mut self) -> ParseResult<'a, Option<&'a [u8]>> {
+        self.try_parse(|p| {
+            p.parse_literal(b"$")?;
+
+            let result = if matches!(p.peek(), Some(b'-')) {
+                p.parse_literal(b"-1")?;
+                None
+            } else {
+                let len = p.parse_len()?;
+                let data = p.parse_bytes(len)?;
+                Some(data)
+            };
+
+            p.parse_crlf()?;
+            Ok(result)
+        })
+    }
+
+    pub fn parse_string(&mut self) -> ParseResult<'a, Option<&'a [u8]>> {
+        match self.peek() {
+            Some(b'+') => self.parse_simple_string().map(Some),
+            Some(b'$') => self.parse_bulk_string(),
+            Some(c) => Err(ParseError::UnexpectedCharacter {
+                expected: "\"+\" or \"$\"",
+                found: c,
+            }),
+            None => Err(ParseError::Incomplete),
+        }
+    }
+
+    pub fn parse_error(&mut self) -> ParseResult<'a, &'a [u8]> {
+        self.try_parse(|p| {
+            p.parse_literal(b"-")?;
+            p.parse_delimited_text()
+        })
+    }
+
+    pub fn parse_literal_u64(&mut self) -> ParseResult<'a, u64> {
+        self.try_parse(|p| {
+            p.parse_literal(b":")?;
+            let text = p.parse_delimited_text()?;
+            let text = std::str::from_utf8(text).map_err(|_| ParseError::invalid_number(text))?;
+            text.parse()
+                .map_err(|_| match <i64 as FromStr>::from_str(text) {
+                    Ok(_) => ParseError::unexpected_negative_number(text.as_bytes()),
+                    Err(_) => ParseError::invalid_number(text.as_bytes()),
+                })
+        })
+    }
+
+    pub fn parse_literal_i64(&mut self) -> ParseResult<'a, i64> {
+        self.try_parse(|p| {
+            p.parse_literal(b":")?;
+            let text = p.parse_delimited_text()?;
+            std::str::from_utf8(text)
+                .map_err(|_| ParseError::invalid_number(text))?
+                .parse()
+                .map_err(|_| ParseError::invalid_number(text))
+        })
+    }
+
+    pub fn parse_array<'p>(&'p mut self) -> ParseResult<'a, Option<ArrayParser<'a, 'p>>> {
+        let len = self.try_parse(|p| {
+            p.parse_literal(b"*")?;
+
+            Ok(match p.peek() {
+                Some(b'-') => {
+                    p.parse_literal(b"-1")?;
+                    p.parse_crlf()?;
+                    None
+                }
+                _ => Some(p.parse_len()?),
+            })
+        })?;
+
+        Ok(len.map(|len| ArrayParser {
+            parser: self,
+            remaining: len,
+        }))
+    }
+
+    pub fn parse_any<V: Visitor<'a>>(&mut self, visitor: V) -> ParseResult<'a, V::Output> {
```

Comment:
Who's this intended for? Not seeing it used within this PR.

Reply:
It was originally used as the base of a `parse_message` type thing but that went away once I introduced `CommandParser`. I think it might be worth keeping in in case someone wants to use this parser for a resp client at some point (e.g. for tests) but I'm not too attached to in and can remove it if you think that's better.